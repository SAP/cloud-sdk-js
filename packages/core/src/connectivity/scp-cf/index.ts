export * from './jwt';
export * from './cache';
export * from './client-credentials-token';
export * from './client-credentials-token-cache';
export * from './connectivity-service';
export * from './connectivity-service-types';
export * from './destination';
export * from './environment-accessor';
export * from './environment-accessor-types';
export * from './resilience-options';
export * from './token-accessor';
export * from './xsuaa-service';
export * from './xsuaa-service-types';
export * from './protocol';
export * from './authorization-header';
export * from './get-protocol';
export * from './header-builder-for-destination';
export {
  mappingTenantFields,
  Tenant,
  tenantFromJwt,
  RegisteredJWTClaimsTenant,
  isIdenticalTenant
} from './tenant';
export {
  RegisteredJWTClaimsUser,
  Scope,
  mappingUserFields,
  userFromJwt,
  customAttributes
} from './user';
