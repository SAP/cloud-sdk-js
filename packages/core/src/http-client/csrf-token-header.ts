import {
  createLogger,
  ErrorWithCause,
  pickIgnoreCase,
  pickNonNullish,
  pickValueIgnoreCase
} from '@sap-cloud-sdk/util';
import { AxiosError } from 'axios';
import { removeTrailingSlashes } from '../odata-common/remove-slashes';
import {
  Destination,
  DestinationNameAndJwt
} from '../connectivity/scp-cf/destination';
import { executeHttpRequest } from '../http-client';
import { HttpRequestConfig } from './http-client-types';

const logger = createLogger({
  package: 'core',
  messageContext: 'csrf-token-header'
});

/**
 * Get CSRF token and cookies for a destination and request configuration. The CSRF token and cookies will be retrieved based on the url of the destination and the custom configuration given by the `requestConfig`.
 * If there is a relative url in the `requestConfig` it will be appended to the destination's url, an absolute url overwrites the destination related url.
 * @param destination The destination to get the headers from
 * @param requestConfig An http request configuration containing additional information about the request, like url or headers
 * @returns A promise to an object containing the CSRF related headers
 */
export async function buildCsrfHeaders<T extends HttpRequestConfig>(
  destination: Destination | DestinationNameAndJwt,
  requestConfig: Partial<T>
): Promise<Record<string, any>> {
  const csrfHeaders = await makeCsrfRequest(destination, requestConfig);
  validateCsrfTokenResponse(csrfHeaders);
  return pickNonNullish({
    ...pickIgnoreCase(csrfHeaders, 'x-csrf-token'),
    cookie: buildCookieHeaderValue(
      pickValueIgnoreCase(csrfHeaders, 'set-cookie')
    )
  });
}

function makeCsrfRequest<T extends HttpRequestConfig>(
  destination: Destination | DestinationNameAndJwt,
  requestConfig: Partial<T>
): Promise<Record<string, any>> {
  const fetchHeader = !pickValueIgnoreCase(
    requestConfig.headers,
    'x-csrf-token'
  ) && { 'x-csrf-token': 'Fetch' };
  const axiosConfig: HttpRequestConfig = {
    method: 'get',
    ...requestConfig,
    headers: {
      ...fetchHeader,
      ...requestConfig.headers
    },
    url: requestConfig.url
  };

  // The S/4 does a redirect if the CSRF token is fetched in case the '/' is not in the URL.
  // TODO: remove once https://github.com/axios/axios/issues/3369 is really fixed. Issue is closed but problem stays.
  const requestConfigWithTrailingSlash = appendSlash(axiosConfig);
  return executeHttpRequest(destination, requestConfigWithTrailingSlash)
    .then(response => response.headers)
    .catch(error1 => {
      if (hasCsrfToken(error1)) {
        return getResponseHeadersFromAxiosError(error1);
      }
      logger.warn(
        new ErrorWithCause(
          `First attempt to fetch CSRF token failed with the url: ${requestConfigWithTrailingSlash.url}. Retrying without trailing slash.`,
          error1
        )
      );
      const requestConfigWithOutTrailingSlash = removeSlash(axiosConfig);
      return executeHttpRequest(destination, requestConfigWithOutTrailingSlash)
        .then(response => response.headers)
        .catch(error2 => {
          if (hasCsrfToken(error2)) {
            return getResponseHeadersFromAxiosError(error2);
          }
          logger.warn(
            new ErrorWithCause(
              `Second attempt to fetch CSRF token failed with the url: ${requestConfigWithOutTrailingSlash.url}. No CSRF token fetched.`,
              error2
            )
          );
          // todo suggest to disable csrf token handling when the api is implemented
          return {};
        });
    });
}

// In general, the valid csrf token head can also be found in a non-2xx request.
// For a non-2xx response, axios will throw an error so we have to check whether the error contains the token.
function hasCsrfToken(error: any): boolean {
  return error.isAxiosError && hasCsrfTokenInAxiosError(error);
}

function hasCsrfTokenInAxiosError(error: AxiosError): boolean {
  return error.response?.headers?.['x-csrf-token'];
}

function getResponseHeadersFromAxiosError(error: AxiosError) {
  return error.response!.headers;
}

function appendSlash(requestConfig: HttpRequestConfig): HttpRequestConfig {
  if (!requestConfig.url!.endsWith('/')) {
    requestConfig.url = `${requestConfig.url}/`;
  }
  return requestConfig;
}

function removeSlash(requestConfig: HttpRequestConfig): HttpRequestConfig {
  if (requestConfig.url!.endsWith('/')) {
    requestConfig.url = removeTrailingSlashes(requestConfig.url!);
  }
  return requestConfig;
}

function validateCsrfTokenResponse(responseHeaders: Record<string, any>) {
  if (!responseHeaders['x-csrf-token']) {
    logger.warn(
      'Destination did not return a CSRF token. This may cause a failure when sending the OData request.'
    );
  }

  if (!responseHeaders['set-cookie']) {
    logger.warn('CSRF header response does not include cookies.');
  }

  return responseHeaders;
}

function buildCookieHeaderValue(cookies?: string[]): string | undefined {
  if (cookies && cookies.length) {
    return cookies.map((cookie: string) => cookie.split(';')[0]).join(';');
  }
}
