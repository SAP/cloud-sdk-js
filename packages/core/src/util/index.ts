export * from './jwt';
export * from './name-converter';
export * from './properties-util';
export * from './proxy-util';
export * from './remove-slashes';
