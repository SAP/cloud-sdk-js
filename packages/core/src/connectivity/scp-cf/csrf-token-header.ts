import {
  createLogger,
  ErrorWithCause,
  pickIgnoreCase,
  pickNonNullish,
  pickValueIgnoreCase
} from '@sap-cloud-sdk/util';
import { AxiosError } from 'axios';
import { HttpRequestConfig, executeHttpRequest } from '../../http-client';
import { Destination, DestinationNameAndJwt } from '../scp-cf';
import { removeTrailingSlashes } from '../../odata-common/remove-slashes';

const logger = createLogger({
  package: 'core',
  messageContext: 'csrf-token-header'
});

/**
 * Get CSRF token and cookies for a destination and request configuration. The CSRF token and cookies will be retrieved based on the url of the destination and the custom configuration given by the `requestConfig`.
 * If there is a relative url in the `requestConfig` it will be appended to the destination's url, an absolute url overwrites the destination related url.
 * @param destination The destination to get the headers from
 * @param requestConfig An http request configuration containing additional information about the request, like url or headers
 * @returns A promise to an object containing the CSRF related headers
 */
export async function buildCsrfHeaders<T extends HttpRequestConfig>(
  destination: Destination | DestinationNameAndJwt,
  requestConfig: Partial<T>
): Promise<Record<string, any>> {
  const csrfHeaders = await makeCsrfRequest(destination, requestConfig);
  validateCsrfTokenResponse(csrfHeaders);
  return pickNonNullish({
    ...pickIgnoreCase(csrfHeaders, 'x-csrf-token'),
    cookie: buildCookieHeaderValue(
      pickValueIgnoreCase(csrfHeaders, 'set-cookie')
    )
  });
}

function makeCsrfRequest<T extends HttpRequestConfig>(
  destination: Destination | DestinationNameAndJwt,
  requestConfig: Partial<T>
): Promise<Record<string, any>> {
  const fetchHeader = !pickValueIgnoreCase(
    requestConfig.headers,
    'x-csrf-token'
  ) && { 'x-csrf-token': 'Fetch' };
  const axiosConfig: HttpRequestConfig = {
    method: 'get',
    ...requestConfig,
    headers: {
      ...fetchHeader,
      ...requestConfig.headers
    },
    url: requestConfig.url
  };

  // The S/4 does a redirect if the CSRF token is fetched in case the '/' is not in the URL.
  // TODO: remove once https://github.com/axios/axios/issues/3369 is really fixed. Issue is closed but problem stays.
  return executeHttpRequest(destination, appendSlash(axiosConfig))
    .then(response => response.headers)
    .catch(e1 => {
      logger.error(
        new ErrorWithCause(
          'Initial try to fetch CSRF token failed - retry without slash at ',
          e1
        )
      );
      return executeHttpRequest(destination, removeSlash(axiosConfig))
        .then(response => response.headers)
        .catch(e2 => {
          throw new ErrorWithCause(
            'Also second try to fetch SRF token failed - No CSRF token fetched.',
            e2
          );
        });
    });
}

function appendSlash(requestConfig: HttpRequestConfig): HttpRequestConfig {
  if (!requestConfig.url!.endsWith('/')) {
    requestConfig.url = `${requestConfig.url}/`;
  }
  return requestConfig;
}

function removeSlash(requestConfig: HttpRequestConfig): HttpRequestConfig {
  if (requestConfig.url!.endsWith('/')) {
    requestConfig.url = removeTrailingSlashes(requestConfig.url!);
  }
  return requestConfig;
}

function axiosWorkaround<T extends HttpRequestConfig>(
  error: AxiosError,
  axiosConfig: Partial<T>,
  destination: Destination | DestinationNameAndJwt
) {
  if (error.request?._isRedirect && error.request?._options?.path) {
    logger.warn(
      'Csrf fetch was redirected and failed. This might be a bug in the underlying request library (https://github.com/axios/axios/issues/3369).\nRetrying with full configuration.'
    );
    return makeCsrfRequest(destination, {
      ...axiosConfig,
      url: error.request._options.path
    });
  }
}

function validateCsrfTokenResponse(responseHeaders: Record<string, any>) {
  if (!responseHeaders['x-csrf-token']) {
    logger.warn(
      'Destination did not return a CSRF token. This may cause a failure when sending the OData request.'
    );
  }

  if (!responseHeaders['set-cookie']) {
    logger.warn('CSRF header response does not include cookies.');
  }

  return responseHeaders;
}

function buildCookieHeaderValue(cookies?: string[]): string | undefined {
  if (cookies && cookies.length) {
    return cookies.map((cookie: string) => cookie.split(';')[0]).join(';');
  }
}
