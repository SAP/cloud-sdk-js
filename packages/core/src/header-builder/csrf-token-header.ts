import { createLogger, errorWithCause } from '@sap-cloud-sdk/util';
import { AxiosError } from 'axios';
import { HttpRequestConfig, executeHttpRequest } from '../http-client';
import { Destination, DestinationNameAndJwt } from '../scp-cf';
import { filterNullishValues, getHeader, getHeaderValue } from './header-util';

const logger = createLogger({
  package: 'core',
  messageContext: 'csrf-token-header'
});

/**
 * Get CSRF token and cookies for a destination and request configuration. The CSRF token and cookies will be retrieved based on the url of the destination and the custom configuration given by the `requestConfig`.
 * If there is a relative url in the `requestConfig` it will be appended to the destination's url, an absolute url overwrites the destination related url.
 * @param destination The destination to get the headers from
 * @param requestConfig An http request configuration containing additional information about the request, like url or headers
 * @returns A promise to an object containing the CSRF related headers
 */
export async function buildCsrfHeaders<T extends HttpRequestConfig>(
  destination: Destination | DestinationNameAndJwt,
  requestConfig: Partial<T>
): Promise<Record<string, string>> {
  const csrfHeaders = await makeCsrfRequest(destination, requestConfig);
  validateCsrfTokenResponse(csrfHeaders);
  return filterNullishValues({
    ...getHeader('x-csrf-token', csrfHeaders),
    cookie: buildCookieHeaderValue(getHeaderValue('set-cookie', csrfHeaders))
  });
}

function makeCsrfRequest<T extends HttpRequestConfig>(
  destination: Destination | DestinationNameAndJwt,
  requestConfig: Partial<T>
): Promise<Record<string, any>> {
  const fetchHeader = !getHeaderValue(
    'x-csrf-token',
    requestConfig.headers
  ) && { 'x-csrf-token': 'Fetch' };
  const axiosConfig: HttpRequestConfig = {
    method: 'get',
    ...requestConfig,
    headers: {
      ...fetchHeader,
      ...requestConfig.headers
    },
    url: requestConfig.url
  };

  return executeHttpRequest(destination, axiosConfig)
    .then(response => response.headers)
    .catch(error => {
      if (!error.response) {
        // TODO: remove once https://github.com/axios/axios/issues/3369 is fixed
        const retry = axiosWorkaround(error, requestConfig, destination);
        if (retry) {
          return retry;
        }
        throw errorWithCause('Csrf fetch failed.', error);
      }
      return error.response.headers;
    });
}

function axiosWorkaround<T extends HttpRequestConfig>(
  error: AxiosError,
  axiosConfig: Partial<T>,
  destination: Destination | DestinationNameAndJwt
) {
  if (error.request?._isRedirect && error.request?._options?.path) {
    logger.warn(
      'Csrf fetch was redirected and failed. This might be a bug in the underlying request library (https://github.com/axios/axios/issues/3369).\nRetrying with full configuration.'
    );
    return makeCsrfRequest(destination, {
      ...axiosConfig,
      url: error.request._options.path
    });
  }
}

function validateCsrfTokenResponse(responseHeaders: Record<string, any>) {
  if (!responseHeaders['x-csrf-token']) {
    logger.warn(
      'Destination did not return a CSRF token. This may cause a failure when sending the OData request.'
    );
  }

  if (!responseHeaders['set-cookie']) {
    logger.warn('CSRF header response does not include cookies.');
  }

  return responseHeaders;
}

function buildCookieHeaderValue(cookies?: string[]): string | undefined {
  if (cookies && cookies.length) {
    return cookies.map((cookie: string) => cookie.split(';')[0]).join(';');
  }
}
