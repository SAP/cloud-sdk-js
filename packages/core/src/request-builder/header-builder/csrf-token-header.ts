/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */

import { createLogger, MapType } from '@sap-cloud-sdk/util';
import { HttpRequestConfig, executeHttpRequest } from '../../http-client';
import { ODataRequest, ODataRequestConfig } from '../request';
import { Destination, DestinationNameAndJwt } from '../../scp-cf';
import { filterNullishValues, getHeader, getHeaderValue } from './headers-util';

const logger = createLogger({
  package: 'core',
  messageContext: 'csrf-token-header'
});

/**
 * Get CSRF token and cookies for a destination and request configuration. The CSRF token and cookies will be retrieved based on the url of the destination and the custom configuration given by the `requestConfig`.
 * If there is a relative url in the `requestConfig` it will be appended to the destination's url, an absolute url overwrites the destination related url.
 * @param destination The destination to get the headers from
 * @param requestConfig An http request configuration containing additional information about the request, like url or headers
 * @returns A promise to an object containing the CSRF related headers
 */
export async function buildCsrfHeaders<T extends HttpRequestConfig>(
  destination: Destination | DestinationNameAndJwt,
  requestConfig: Partial<T>
): Promise<MapType<string>> {
  const csrfHeaders = await makeCsrfRequest(destination, requestConfig);
  validateCsrfTokenResponse(csrfHeaders);
  return filterNullishValues({
    ...getHeader('x-csrf-token', csrfHeaders),
    cookie: buildCookieHeaderValue(getHeaderValue('set-cookie', csrfHeaders))
  });
}

function makeCsrfRequest<T extends HttpRequestConfig>(
  destination: Destination | DestinationNameAndJwt,
  requestConfig: Partial<T>
): Promise<MapType<any>> {
  const fetchHeader = !getHeaderValue(
    'x-csrf-token',
    requestConfig.headers
  ) && { 'x-csrf-token': 'Fetch' };
  const axiosConfig: HttpRequestConfig = {
    method: 'get',
    ...requestConfig,
    headers: {
      ...fetchHeader,
      ...requestConfig.headers
    },
    url: requestConfig.url
  };

  return executeHttpRequest(destination, axiosConfig)
    .then(response => response.headers)
    .catch(error => {
      if (!error.response) {
        throw new Error('The error response is undefined.');
      }
      return error.response.headers;
    });
}

function validateCsrfTokenResponse(responseHeaders: MapType<any>) {
  if (!responseHeaders['x-csrf-token']) {
    logger.warn(
      'Destination did not return a CSRF token. This may cause a failure when sending the OData request.'
    );
  }

  if (!responseHeaders['set-cookie']) {
    logger.warn('CSRF header response does not include cookies.');
  }

  return responseHeaders;
}

function buildCookieHeaderValue(cookies?: string[]): string | undefined {
  if (cookies && cookies.length) {
    return cookies.map((cookie: string) => cookie.split(';')[0]).join(';');
  }
}

/**
 * @deprecated Since v1.20.0, use [[buildCsrfHeaders]] instead.
 *
 * Add CSRF token and cookies for a request to destination related headers.
 * @param request The request to get CSRF headers for.
 * @param headers Destination related headers to include in the request.
 * @returns A promise to an object containing the CSRF related headers
 */
export async function addCsrfTokenAndCookies<
  RequestT extends ODataRequestConfig
>(
  request: ODataRequest<RequestT>,
  headers: MapType<string>
): Promise<MapType<string>> {
  if (!request.destination) {
    throw Error('The request destination is undefined.');
  }
  if (
    request.config.method === 'get' ||
    Object.keys(headers).includes('x-csrf-token')
  ) {
    return headers;
  }
  return {
    ...(await buildCsrfHeaders(request.destination, {
      headers,
      url: request.relativeServiceUrl()
    })),
    ...headers
  };
}
