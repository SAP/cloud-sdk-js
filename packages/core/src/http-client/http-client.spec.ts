import https from 'https';
import nock from 'nock';
import { createLogger } from '@sap-cloud-sdk/util';
import axios from 'axios';
import { Destination, Protocol } from '../connectivity';
import * as csrfHeaders from './csrf-token-header';
import {
  DestinationHttpRequestConfig,
  HttpMethod,
  HttpRequest,
  HttpRequestConfig,
  HttpResponse
} from './http-client-types';
import {
  addDestinationToRequestConfig,
  buildHttpRequest,
  executeHttpRequest,
  shouldHandleCsrfToken
} from './http-client';

describe('generic http client', () => {
  const httpsDestination: Destination = {
    name: 'httpsDestination',
    url: 'https://example.com',
    authentication: 'BasicAuthentication',
    username: 'USERNAME',
    password: 'PASSWORD',
    sapClient: '001'
  };

  const proxyAuthorization = 'youmaypass';

  const proxyDestination: Destination = {
    name: 'proxyDestination',
    url: 'http://example.com',
    authentication: 'BasicAuthentication',
    proxyType: 'Internet',
    username: 'USERNAME',
    password: 'PASSWORD',
    sapClient: '001',
    proxyConfiguration: {
      headers: {
        'Proxy-Authorization': proxyAuthorization
      },
      host: 'proxy.host',
      port: 1234,
      protocol: Protocol.HTTP
    }
  };

  describe('buildHttpRequest', () => {
    it('provides a baseURL, headers, and either an httpAgent or an httpsAgent', async () => {
      const expectedHttps: DestinationHttpRequestConfig = {
        baseURL: 'https://example.com',
        headers: {
          authorization: 'Basic VVNFUk5BTUU6UEFTU1dPUkQ=',
          'sap-client': '001'
        }
      };

      const actualHttps = await buildHttpRequest(httpsDestination);

      expect(actualHttps).toMatchObject(expectedHttps);
      expect(actualHttps.httpsAgent).toBeDefined();

      const expectedProxy: DestinationHttpRequestConfig = {
        baseURL: 'http://example.com',
        headers: {
          authorization: 'Basic VVNFUk5BTUU6UEFTU1dPUkQ=',
          'sap-client': '001',
          'Proxy-Authorization': proxyAuthorization
        }
      };

      const actualProxy = await buildHttpRequest(proxyDestination);

      expect(actualProxy).toMatchObject(expectedProxy);
      expect(actualProxy.httpAgent).toBeDefined();
    });

    it('warn when custom headers are used', async () => {
      const logger = createLogger({
        package: 'core',
        messageContext: 'http-client'
      });
      const warnSpy = jest.spyOn(logger, 'warn');

      await buildHttpRequest(httpsDestination, {
        authorization: 'abc',
        'sap-client': '001',
        'SAP-Connectivity-SCC-Location_ID': 'efg'
      });
      expect(warnSpy).toBeCalledWith(
        `The following custom headers will overwrite headers created by the SDK:
  - "authorization"
  - "sap-client"
  - "SAP-Connectivity-SCC-Location_ID"`
      );
    });

    it('throws useful error messages when finding the destination fails', async () => {
      await expect(
        buildHttpRequest({ destinationName: 'does not exist' })
      ).rejects.toThrowErrorMatchingInlineSnapshot(
        '"Failed to load destination."'
      );
    });

    it('throws useful error messages when building headers fails', async () => {
      await expect(
        buildHttpRequest({
          url: 'https://example.com',
          authentication: 'BasicAuthentication'
        })
      ).rejects.toThrowErrorMatchingInlineSnapshot(
        '"Failed to build headers."'
      );
    });

    it('defaults to NoAuthentication/does not throw if no credentials are provided', async () => {
      await expect(
        buildHttpRequest({ url: 'https://example.com' })
      ).resolves.not.toThrow();
    });
  });

  describe('addDestinationToRequestConfig', () => {
    afterEach(() => {
      jest.clearAllMocks();
    });

    it('merges given headers with provided headers (but gives priority to custom headers)', async () => {
      const expected = {
        a: '1',
        b: '2',
        baseURL: 'https://example.com',
        headers: {
          a: '1',
          authorization: 'CUSTOM',
          'sap-client': '001'
        }
      };

      const actual = await addDestinationToRequestConfig(httpsDestination, {
        method: 'get',
        a: '1',
        b: '2',
        headers: {
          a: '1',
          authorization: 'CUSTOM'
        }
      });

      expect(actual).toMatchObject(expected);
      expect(actual.httpsAgent).toBeDefined();
    });

    it('overwrites baseURL and either httpAgent or httpsAgent', async () => {
      const expected = {
        baseURL: 'https://custom.example.com',
        headers: {
          a: '1',
          authorization: 'CUSTOM',
          'sap-client': '001'
        },
        httpsAgent: 'custom-agent'
      };

      const actual = await addDestinationToRequestConfig(httpsDestination, {
        method: 'get',
        baseURL: 'https://custom.example.com',
        headers: {
          a: '1',
          authorization: 'CUSTOM'
        },
        httpsAgent: 'custom-agent'
      });

      expect(actual).toMatchObject(expected);
    });

    it('allows usage with e.g. axios.request', async () => {
      nock('https://example.com', {
        reqheaders: {
          authorization: 'Basic VVNFUk5BTUU6UEFTU1dPUkQ=',
          'sap-client': '001'
        }
      })
        .get('/api/entity')
        .query({
          a: 'a',
          b: 'b'
        })
        .reply(200);

      const request = addDestinationToRequestConfig(httpsDestination, {
        method: 'get',
        url: '/api/entity',
        headers: {},
        params: {
          a: 'a',
          b: 'b'
        }
      }).then(conf => axios.request(conf));
      await expect(request).resolves.not.toThrow();
    });
  });

  describe('executeHttpRequest', () => {
    afterEach(() => {
      jest.restoreAllMocks();
    });

    // The base-agent dependency coming in via the http-proxy-agent did mess with the node https.
    // Since version 3.0.0 the issue is fixed and the test ensures this.
    it('should not break the plain node http client', async () => {
      nock('https://example.com').get(/.*/).reply(200);

      return new Promise((resolve, reject) =>
        https.get('https://example.com', response =>
          response.statusCode === 200 ? resolve(200) : reject()
        )
      );
    });

    it('takes a generic HTTP request and executes it', async () => {
      nock('https://example.com', {
        reqheaders: {
          authorization: 'Basic VVNFUk5BTUU6UEFTU1dPUkQ=',
          'sap-client': '001'
        }
      })
        .get('/api/entity')
        .query({
          a: 'a',
          b: 'b'
        })
        .reply(200, { res: 'ult' }, { sharp: 'header' });

      const config = {
        method: HttpMethod.GET,
        url: '/api/entity',
        params: {
          a: 'a',
          b: 'b'
        }
      };

      const response = await executeHttpRequest(httpsDestination, config);
      expect(response.data.res).toBe('ult');
      expect(response.status).toBe(200);
      expect(response.headers).toMatchObject({ sharp: 'header' });
    });

    it('also works also in more complex cases in more complex cases', async () => {
      nock('https://custom.example.com', {
        reqheaders: {
          'content-type': 'application/json',
          accept: 'application/json',
          'x-csrf-token': 'Fetch',
          authorization: 'custom-auth-header',
          'sap-client': '001'
        }
      })
        .post('/api/entity', {
          a: 1,
          b: 2,
          c: 3
        })
        .reply(200);

      const config: HttpRequest = {
        baseURL: 'https://custom.example.com',
        method: HttpMethod.POST,
        url: '/api/entity',
        headers: {
          'content-type': 'application/json',
          accept: 'application/json',
          'x-csrf-token': 'Fetch',
          authorization: 'custom-auth-header'
        },
        data: {
          a: 1,
          b: 2,
          c: 3
        }
      };

      await expect(
        executeHttpRequest(httpsDestination, config)
      ).resolves.not.toThrow();
    });

    it('propagates errors to the caller', async () => {
      nock('https://example.com', {
        reqheaders: {
          authorization: 'Basic VVNFUk5BTUU6UEFTU1dPUkQ=',
          'sap-client': '001'
        }
      })
        .get('/api/entity')
        .query({
          a: 'a',
          b: 'b'
        })
        .reply(500, { res: 'NOPE' }, { sharp: 'header' });

      await expect(
        executeHttpRequest(httpsDestination, {
          method: 'get',
          url: '/api/entity',
          params: {
            a: 'a',
            b: 'b'
          }
        })
      ).rejects.toThrowErrorMatchingInlineSnapshot(
        '"Request failed with status code 500"'
      );
    });

    it('fetches csrf token headers when fetchCsrfToken is true', async () => {
      const csrfToken = 'some-csrf-token';
      nock('https://example.com', {
        reqheaders: {
          'x-csrf-token': 'Fetch',
          'content-type': 'application/json',
          accept: 'application/json',
          authorization: 'custom-auth-header',
          'sap-client': '001'
        }
      })
        .head('/api/entity')
        .reply(200, {}, { 'x-csrf-token': csrfToken });

      nock('https://example.com', {
        reqheaders: {
          'x-csrf-token': csrfToken,
          'content-type': 'application/json',
          accept: 'application/json',
          authorization: 'custom-auth-header',
          'sap-client': '001'
        }
      })
        .post('/api/entity', {
          a: 1
        })
        .reply(200);

      const config: HttpRequest = {
        baseURL: 'https://example.com',
        method: HttpMethod.POST,
        url: '/api/entity',
        headers: {
          authorization: 'custom-auth-header',
          'content-type': 'application/json',
          accept: 'application/json'
        },
        data: {
          a: 1
        }
      };

      await expect(
        executeHttpRequest(httpsDestination, config, { fetchCsrfToken: true })
      ).resolves.not.toThrow();
    });

    it('fetches csrf token headers even the status of the response is not 2xx', async () => {
      const csrfToken = 'some-csrf-token';
      nock('https://example.com', {
        reqheaders: {
          'x-csrf-token': 'Fetch',
          'content-type': 'application/json',
          accept: 'application/json',
          authorization: 'custom-auth-header',
          'sap-client': '001'
        }
      })
        .head('/api/entity')
        .reply(405, {}, { 'x-csrf-token': csrfToken });

      nock('https://example.com', {
        reqheaders: {
          'x-csrf-token': csrfToken,
          'content-type': 'application/json',
          accept: 'application/json',
          authorization: 'custom-auth-header',
          'sap-client': '001'
        }
      })
        .post('/api/entity', {
          a: 1
        })
        .reply(200);

      const config: HttpRequest = {
        baseURL: 'https://example.com',
        method: HttpMethod.POST,
        url: '/api/entity',
        headers: {
          authorization: 'custom-auth-header',
          'content-type': 'application/json',
          accept: 'application/json'
        },
        data: {
          a: 1
        }
      };

      await expect(
        executeHttpRequest(httpsDestination, config, { fetchCsrfToken: true })
      ).resolves.not.toThrow();
    });

    it('should not fetch csrf token headers when fetchCsrfToken is true and the the request contains the token in the header', async () => {
      const csrfToken = 'some-csrf-token';

      nock('https://example.com', {
        reqheaders: {
          'x-csrf-token': csrfToken,
          'content-type': 'application/json',
          accept: 'application/json',
          authorization: 'custom-auth-header',
          'sap-client': '001'
        }
      })
        .post('/api/entity', {
          a: 1
        })
        .reply(200);

      const config: HttpRequest = {
        baseURL: 'https://example.com',
        method: HttpMethod.POST,
        url: '/api/entity',
        headers: {
          authorization: 'custom-auth-header',
          'content-type': 'application/json',
          accept: 'application/json',
          'x-csrf-token': csrfToken
        },
        data: {
          a: 1
        }
      };

      jest.spyOn(csrfHeaders, 'buildCsrfHeaders');
      await executeHttpRequest(httpsDestination, config, {
        fetchCsrfToken: true
      });
      expect(csrfHeaders.buildCsrfHeaders).not.toHaveBeenCalled();
    });

    it('should apply http(s)Agent to both CSRF token request and the real request', async () => {
      const csrfTokenResponse: HttpResponse = {
        headers: { 'x-csrf-token': 'token' }
      } as HttpResponse;
      const spy = jest
        .spyOn(axios, 'request')
        .mockImplementationOnce(async () => csrfTokenResponse);
      jest
        .spyOn(axios, 'request')
        .mockImplementationOnce(async () => 'post response');

      const httpsAgentOption = { ca: 'CA' };
      const config: HttpRequest = {
        baseURL: 'https://www.example.com',
        url: 'api',
        method: 'post',
        headers: {},
        httpsAgent: new https.Agent(httpsAgentOption)
      };

      await executeHttpRequest(httpsDestination, config, {
        fetchCsrfToken: true
      });
      expect(spy).toHaveBeenNthCalledWith(
        2,
        expect.objectContaining({
          httpsAgent: expect.objectContaining({
            options: expect.objectContaining(httpsAgentOption)
          })
        })
      );
    });

    it('includes the default axios config in request', async () => {
      const destination: Destination = { url: 'https://destinationUrl' };
      const requestSpy = jest.spyOn(axios, 'request').mockResolvedValue(true);
      await executeHttpRequest(destination, { method: 'get' });
      expect(requestSpy).toHaveBeenCalledWith(
        expect.objectContaining({
          proxy: false,
          httpsAgent: expect.anything(),
          httpAgent: expect.anything()
        })
      );

      requestSpy.mockRestore();
    });

    it('overwrites the default axios config with destination related request config', async () => {
      const destination: Destination = {
        url: 'https://destinationUrl',
        proxyConfiguration: {
          host: 'dummy',
          port: 1234,
          protocol: Protocol.HTTP
        }
      };
      const requestSpy = jest.spyOn(axios, 'request').mockResolvedValue(true);
      await expect(
        executeHttpRequest(destination, { method: 'get' })
      ).resolves.not.toThrow();
      expect(requestSpy).toHaveBeenCalledWith(
        expect.objectContaining({
          proxy: false,
          httpsAgent: expect.objectContaining({
            proxy: expect.objectContaining({ port: 1234 })
          })
        })
      );
    });

    it('overwrites destination related request config with the explicit one', async () => {
      const destination: Destination = { url: 'https://destinationUrl' };
      const requestSpy = jest.spyOn(axios, 'request').mockResolvedValue(true);
      await expect(
        executeHttpRequest(destination, { method: 'post' })
      ).resolves.not.toThrow();

      expect(requestSpy).toHaveBeenCalledWith(
        expect.objectContaining({
          method: 'post'
        })
      );
    });
  });

  describe('shouldHandleCsrfToken', () => {
    it('should not handle csrf token for get request', () => {
      const request = { method: 'get' } as HttpRequestConfig;
      const options = { fetchCsrfToken: true };
      expect(shouldHandleCsrfToken(request, options)).toEqual(false);
    });

    it('should not handle csrf token when fetchCsrfToken is false', () => {
      const request = { method: 'post' } as HttpRequestConfig;
      const options = { fetchCsrfToken: false };
      expect(shouldHandleCsrfToken(request, options)).toEqual(false);
    });

    it('should handle csrf token for non-get request when fetchCsrfToken is true', () => {
      const request = { method: 'patch' } as HttpRequestConfig;
      const options = { fetchCsrfToken: true };
      expect(shouldHandleCsrfToken(request, options)).toEqual(true);
    });
  });
});
