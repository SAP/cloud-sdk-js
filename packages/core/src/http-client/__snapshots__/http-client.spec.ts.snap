// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generic http client buildHttpRequest throws useful error messages when building headers fails 1`] = `"Failed to build HTTP request for destination: failed to build headers!"`;

exports[`generic http client buildHttpRequest throws useful error messages when finding the destination fails 1`] = `"Failed to build HTTP request for destination: failed to load destination!"`;

exports[`generic http client executeHttpRequest propagates errors to the caller 1`] = `"Request failed with status code 500"`;
